datafusion/core/src/physical_optimizer/pruning.rs

```diff
@@ -2484,10 +2630,466 @@ mod tests {
+                    Some(false),
+                    // container 8 unknown about "foo"
+                    None,
+                ],
+            )
+            .with_contained(
+                "s1",
+                [ScalarValue::from("bar")],
+                [
+                    // containers 0,1,2 known to only contain "bar"
+                    Some(true),
+                    Some(true),
+                    Some(true),
+                    // container 3,4,5 known to not contain "bar"
+                    Some(false),
+                    Some(false),
+                    Some(false),
+                    // container 6,7,8 unknown about "bar"
+                    None,
+                    None,
+                    None,
+                ],
+            )
+            .with_contained(
+                // the way the tests are setup, this data is
+                // consulted if the "foo" and "bar" are being checked at the same time
+                "s1",
+                [ScalarValue::from("foo"), ScalarValue::from("bar")],
+                [
+                    // container 0,1,2 unknown about ("foo, "bar")
+                    None,
+                    None,
+                    None,
+                    // container 3,4,5 known to contain only either "foo" and "bar"
+                    Some(true),
+                    Some(true),
+                    Some(true),
+                    // container 6,7,8  known ro contain  neither "foo" and "bar"
+                    Some(false),
+                    Some(false),
+                    Some(false),
+                ],
+            );
+
+        // s1 = 'foo'
+        prune_with_expr(
+            col("s1").eq(lit("foo")),
+            &schema,
+            &statistics,
+            // rule out containers ('false) where we know foo is not present
+            vec![true, false, true, true, false, true, true, false, true],
+        );
+
+        // s1 = 'bar'
+        prune_with_expr(
+            col("s1").eq(lit("bar")),
+            &schema,
+            &statistics,
+            // rule out containers where we know bar is not present
+            vec![true, true, true, false, false, false, true, true, true],
+        );
+
+        // s1 = 'baz' (unknown value)
+        prune_with_expr(
+            col("s1").eq(lit("baz")),
+            &schema,
+            &statistics,
+            // can't rule out anything
+            vec![true, true, true, true, true, true, true, true, true],
+        );
+
+        // s1 = 'foo' AND s1 = 'bar'
+        prune_with_expr(
+            col("s1").eq(lit("foo")).and(col("s1").eq(lit("bar"))),
+            &schema,
+            &statistics,
+            // logically this predicate can't possibly be true (the column can't
+            // take on both values) but we could rule it out if the stats tell
+            // us that both values are not present
+            vec![true, true, true, true, true, true, true, true, true],
```

Comment:
> but we could rule it out if the stats tell
> // us that both values are not present

Is it possible to have a test container `false` to say `both values are not present`?

Reply:
Yes, you can say that by returning `false` for `contained("s1", {foo, bar})`

However, in this case I think what happens is we end up with two distinct literal guarantees and the container would have to know that a container only had foo AND only had bar, which is logically impossible. 

So in other words, this expression
```
Pruning with expr: s1 != Utf8("foo") AND s2 != Utf8("bar")
```

Generates these guarantees:
```
   Got guarantees: [
    LiteralGuarantee { column: Column { relation: None, name: "s1" }, guarantee: NotIn, literals: {Utf8("foo")} },
    LiteralGuarantee { column: Column { relation: None, name: "s2" }, guarantee: NotIn, literals: {Utf8("bar")} }
  ]
```

I think it would be possiible to do another round of analysis on this and prove this can never be true. I am not sure how important the use case is however.

Reply:
Makes sense.Thanks Andrew